### ISSUE TITLE
Shall we add select through by keymap modification?

### ISSUE THREAD

**Issue by ReinerBforartists** [Bforartists/Bforartists#3004]
Forester 64 has brought up an interesting keymap modification for the case that you want to use select through without code modifications. He inserts a toggle into the chain.

https://devtalk.blender.org/t/decoupling-x-ray-and-limit-selection-to-visible/3498/454

Is this a good idea? Shall we have a look to implement this way into our keymap?

**Comment by Draise14** [comment]
X for Xray?

**Comment by ReinerBforartists** [comment]
X C V  is occupied by the vertice , edge, and face mode. And brushes. This makes it so hard here. We're once more running out of useful hotkeys ...

**Comment by ReinerBforartists** [comment]
The Y key could be an option, with some tweaks. Who nees a hotkey to switch between grease pencil layers?. But in the american keyboard layout this is switched with the Z key. So it's harder to grab there.

![ykey](https://user-images.githubusercontent.com/13919702/168747036-f8f51c13-e0f5-4644-bb9d-247d2db4ef04.jpg)

**Comment by Draise14** [comment]
Yeah, Y is a bit far on most EN/ES keyboards, two of the most common spoken languages. Hmm.. Maybe a modal like CTRL/ALT/SHIFT? 

Man I hate the keymap manager, we really need conflict checkers and an actual keyboard layout with drag and drop assignment..

**Comment by ReinerBforartists** [comment]
No. We need a new keymap manager. But this is behind our scope. Even when we would manage it, we would break addon compatiblity.

**Comment by ReinerBforartists** [comment]
I suggest to put this idea on ice for a while. They are at heavily discussing the issue at blender developer at the moment. Maybe we come closer to an official solution.

**Comment by ReinerBforartists** [comment]
Given that the Blender keymap uses alt z to toggle the x ray mode, i think we are perfectly fine with the Y key here. American users will prefer the toggle x ray mode then, and european users can benefit from select through ...

But first let's catch the bear before we distribute its fur. Investigating ...

**Comment by ReinerBforartists** [comment]
And there runs the bear ... 

I cannot get the box select part to work. But alone toggling the x ray as described would be already a big relief. Press = xray on, release button = xray off. So investigating if we can catch at least this part of it.

I have currently mapped it to shift y. This shortcut is free. Ctrl or alt collides with the modal map of box select in the tool shelf. But we could also think to add an addition to the existing hotkey for x ray. Which is alt z. But z is harder to grab at my european keyboard. Y is occupied by more than just the grease pencil. And when we use the knife tool, then we might run into unexpected behaviour here. Y is in the modal map. So i want to test this for a while before i push it into master.

Ideas?

![togglexray](https://user-images.githubusercontent.com/13919702/175012365-f34e3753-59f0-4af6-8d97-233732f0e373.jpg)


![possibleconflict](https://user-images.githubusercontent.com/13919702/175012380-9d903c45-2bc1-4169-8187-6027025e845c.jpg)

**Comment by ReinerBforartists** [comment]
I think i found a solution that is even better than any of these ideas.

By adding a mouse hotkey to toggle x ray. This is just possible with our keymap though. In Blender and with the Blender keymap you run into heavy conflicts. But here it works just fine since we have rmb mapped to navigation, so no conflict. And it speeds up things in a nice and comfortable way :)

B right mouse for box select. And g right mouse for circle select. We have no hotkey for lasso select though. But b or c right mouse already toggles the two most used methods.

What do you think?

![contexttoggle](https://user-images.githubusercontent.com/13919702/175355274-66034718-a639-41ac-b229-0f20ed779008.jpg)

**Comment by ReinerBforartists** [comment]
Thanks :)

What still bothers me a bit is to have a hotkey for a setting to have bound to a tool.

The culprit is that at an european keyboard the hotkey to toggle x ray is simply very uncomfortable to grab. Since z and y are switched here. So one thing that we could do is to add  the alt y key to toggle x ray. This would make the hotkey equal comfortable at every keyboard layout.

And i think to assign the right mouse functionality to either z or y then, and not to the B key for box select. Since we have lasso and circle select too. And tweak ...

Let's do some experiments in this regards.

**Comment by ReinerBforartists** [comment]
Y would work. And Z too. For all selection methods. And there is no conflict with the knife tool really ( i would nevertheless love when somebody would test this part). Grease pencil has its own separated modes. So this is out of the way.

So question, shall we use Y? Or Z? Z is already connected with the toggle x ray. 

Or shall we add both? As told, german keyboard users are uncomfortable with the hotkey alt z. So i could add alt y, and then y right mouse and z right mouse too.

Or we leave the right mouse idea away, and just add alt y as an alternative. Better to grab ...

Ideas?

**Comment by ReinerBforartists** [comment]
Tested it a few hours now, and the more comfortable way is to add the alternative hotkey alt y to toggle the select through for german keyboard layout. You need to press a hotkey to toggle anyways. So the mouse way wins nothing in this regards really.

The better method would be of course when we could have something like a modal map for the select tool, not dependant of what view you are. Like holding right mouse will select through then. Which is currently simply not to achieve. Or something like in the select through patch. But we had the select through patch already applied, and it did not work so fine. And would add to us the burden to maintain it then.

So i would say let's add alt y and call this task a day then.
